### PR TITLE
NEW: Optional median filtering of probe "hot spots"

### PR DESCRIPTION
**Pull request by a4894z**
This doesn't fix the underlying cause of these hot spots but this should cure the symptoms of the underlying disease for the time being

## Purpose
Sometimes we see these strange "hot spots" in the shared probe modes; we are still unsure of the underlying cause of these hot spots but use of a median filter seems to be effective of removing them when do occur. 

## Approach
I introduce a function that simply takes the absolute value of the shared probe modes, and applies a median filter to each mode. The width of the median window is specified in the .toml file; alternatively we default to NOT using any median filtering.

### Submitter
- [ ] Write a helpfully descriptive pull request title.
- [ ] Organize changes into logically grouped commits with descriptive commit messages.
- [ ] Document all new functions.
- [ ] Click 'details' on the readthedocs check to view the updated docs.
- [ ] Write tests for new functions or explain why they are not needed.
- [ ] Address any complaints from pep8speaks.

### Reviewer
- [x] Actually read all of the code.
- [ ] Run the new code yourself; the included tests should make this easy.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

**Review by carterbox — CHANGES_REQUESTED** [review]

File: src/tike/ptycho/probe.py [hunk 1]
```diff
@@ -870,6 +887,16 @@ def constrain_center_peak(probe):
     probe = stack.reshape(probe.shape)
     return probe
 
+def apply_median_filter_abs_probe( probe, med_filt_px ):
+    """ ABCDEF """
```
[the remark this answers is not in the document]
Comment (a4894z) [reply src/tike/ptycho/probe.py]: hopefully it's ok now

File: src/tike/ptycho/probe.py [hunk 2]
```diff
@@ -138,6 +138,20 @@ class ProbeOptions:
     hard constraint.
     """
 
+    median_filter_abs_probe: bool = False
+    """Binary switch on whether to apply a median filter to absolute value of 
+    each shared probe mode.
+    """
```
Comment (carterbox) [inline src/tike/ptycho/probe.py:144]: Aren't `median_filter_abs_probe` and `median_filter_abs_probe_period` redundant? For example, you could set the filter period to 0 or something large like 2^32-1 to disable it?

File: src/tike/ptycho/probe.py [hunk 3]
```diff
@@ -138,6 +138,20 @@ class ProbeOptions:
...
+
+    median_filter_abs_probe_px: typing.Tuple[float, float] = ( 1.0, 1.0 )
+    """A 2-element tuple with the median filter pixel widths along each dimension.
+    """
+
+    median_filter_abs_probe_period: int = 1
+    """How often (wrt update epochs) we want to apply median filtering to the 
+    absolute value of each shared probe mode.
+    """
```
Comment (carterbox) [inline src/tike/ptycho/probe.py:153]: What happens when median_filter_abs_probe_period and update_period are not the same? For example, if you only update the probe every 5th epoch, but you set the filter period to every 3rd period, then you only filter when the epoch is a multiple of both 5 and 3? Is that intentional?

```
update : 0, 5, 10, [15], 20, 25, [30]
filter : 0, 3, 6, 9, 12, [15], 18, 21, 24, 27, [30]
```

**Review by carterbox — CHANGES_REQUESTED** [review]

File: src/tike/ptycho/probe.py (hunk 2 above)
Comment (a4894z) [reply src/tike/ptycho/probe.py:144]: I've removed `median_filter_abs_probe_period` altogether; see other comment below.

File: src/tike/ptycho/probe.py (hunk 3 above)
Comment (a4894z) [reply src/tike/ptycho/probe.py]: Yeah, this is kind of a mess...

I'm starting to think we should just have `median_filter_abs_probe` as a binary switch and `median_filter_abs_probe_px` as the median filter pixel widths.

I've removed `median_filter_abs_probe_period` altogether as the hot spot suppression needs this to happen every probe update epoch anyway.

Comment (carterbox) [reply src/tike/ptycho/probe.py]: Makes sense, thanks!

**Review by carterbox — APPROVED** [review]

**Comment by carterbox** [comment]
I thought automerge would wait until the tests pass. Oops!

**Pull request merged**